src/gpl/src/mbff.cpp

```diff
@@ -957,116 +1160,374 @@ void MBFF::KMeansDecomp(const std::vector<Flop>& flops,
+    std::vector<Flop> flops_0;
+    std::vector<Flop> flops_1;
+    for (int idx : clks.second) {
+      int setting = GetNumQ(insts_[idx]) - GetNumD(insts_[idx]);
+      if (!setting) {
+        flops_0.push_back(flops_[idx]);
+      } else {
+        flops_1.push_back(flops_[idx]);
+      }
+    }
+    if (flops_0.size()) {
+      ffs.push_back(flops_0);
+    }
+    if (flops_1.size()) {
+      ffs.push_back(flops_1);
+    }
+  }
 }
 
-// ctor
-MBFF::MBFF(int num_flops,
-           int num_paths,
-           const std::vector<odb::Point>& points,
-           const std::vector<Path>& paths,
-           int threads,
-           int knn,
-           int multistart,
-           utl::Logger* log)
+void MBFF::SetTrayNames()
 {
-  flops_.reserve(num_flops);
-  for (int i = 0; i < num_flops; i++) {
-    Flop new_flop{points[i], i, 0.0};
-    flops_.emplace_back(new_flop);
+  for (size_t i = 0; i < 2 * (flops_.size()); i++) {
+    unused_.push_back(i);
   }
+}
 
-  paths_.reserve(num_paths);
-  for (int i = 0; i < num_paths; i++) {
-    paths_.emplace_back(paths[i]);
-    flops_in_path_.insert(paths[i].start_point);
-    flops_in_path_.insert(paths[i].end_point);
+void MBFF::Run(int mx_sz, double alpha, double beta)
+{
+  srand(1);
+  omp_set_num_threads(num_threads_);
+
+  ReadFFs();
+  ReadPaths();
+  ReadLibs();
+  SetTrayNames();
+
+  std::vector<std::vector<Flop>> FFs;
+  SeparateFlops(FFs);
+  int num_chunks = static_cast<int>(FFs.size());
+  double tot_ilp = 0;
+  for (int i = 0; i < num_chunks; i++) {
+    SetVars(FFs[i]);
+    int setting = GetNumQ(insts_[FFs[i].back().idx])
+                  - GetNumD(insts_[FFs[i].back().idx]);
+    SetRatios(setting);
+    tot_ilp += doit(FFs[i], mx_sz, alpha, beta, setting);
   }
+  double tcp_disp = (beta * GetTCPDisplacement());
 
-  slot_disp_x_.reserve(num_flops);
-  slot_disp_y_.reserve(num_flops);
-  for (int i = 0; i < num_flops; i++) {
-    slot_disp_x_.emplace_back(0);
-    slot_disp_y_.emplace_back(0);
+  // round
+  tot_ilp = std::ceil(tot_ilp * multiplier_);
+  tot_ilp /= multiplier_;
+
+  tcp_disp = std::ceil(tcp_disp * multiplier_);
+  tcp_disp /= multiplier_;
+
+  log_->report("Alpha = {}, Beta = {}, max size = {}", alpha, beta, mx_sz);
+  log_->report("Total ILP Cost: {}", tot_ilp);
+  log_->report("Total Timing Critical Path Displacement: {}", tcp_disp);
+  log_->report("Final Objective Value: {}", tot_ilp + tcp_disp);
+}
+
+void MBFF::ReadLibs()
+{
+  for (int i = 0; i < 2; i++) {
+    best_master_[i].resize(num_sizes_, nullptr);
+    tray_area_[i].resize(num_sizes_, std::numeric_limits<double>::max());
+    tray_width_[i].resize(num_sizes_);
+    pin_mappings_[i].resize(num_sizes_);
+
+    slot_to_tray_x_[i].resize(num_sizes_);
+    slot_to_tray_y_[i].resize(num_sizes_);
   }
 
+  for (auto lib : db_->getLibs()) {
+    for (auto master : lib->getMasters()) {
+      odb::dbInst* tmp_tray = odb::dbInst::create(block_, master, "test_tray");
+      if (!IsValidTray(tmp_tray)) {
```

Comment:
Is you pass every type of cell to IsValidTray you can expect IsDPin not to work on most (eg a nand2 will think it has 2 D pins).

Reply:
Does the following pseudocode seem fine? 

IsValidTray(inst):
 slots = 0
 for (seq : sequentials):
   slots += (seq->PortIterator->hasNext())
 return slots > 1
 

Since we scan sequentials (and are essentially checking that it's not an FF), I believe this should work

Reply:
ok